meter/sma.go

```diff
@@ -27,62 +29,67 @@ type SMA struct {
 	mux     *util.Waiter
 	uri     string
 	serial  string
-	iface 	string
+	iface   string
 	values  values
 	powerO  sma.Obis
 	energyO sma.Obis
-	recv    chan sma.Telegram
+
+	device       *sunny.Device
+	updateTicker *time.Ticker
 }
 
 func init() {
 	registry.Add("sma", NewSMAFromConfig)
 }
 
-//go:generate go run ../cmd/tools/decorate.go -f decorateSMA -b api.Meter -t "api.MeterEnergy,TotalEnergy,func() (float64, error)"
+//go:generate go run ../cmd/tools/decorate.go -f decorateSMA -b api.Meter -t "api.MeterEnergy,TotalEnergy,func() (float64, error)" -t "api.Battery,SoC,func() (float64, error)"
 
 // NewSMAFromConfig creates a SMA Meter from generic config
 func NewSMAFromConfig(other map[string]interface{}) (api.Meter, error) {
 	cc := struct {
-		URI, Serial, Interface, Power, Energy string
+		URI, Password, Serial, Interface, Power, Energy string
 	}{}
 
 	if err := util.DecodeOther(other, &cc); err != nil {
 		return nil, err
 	}
 
-	return NewSMA(cc.URI, cc.Serial, cc.Interface, cc.Power, cc.Energy)
+	return NewSMA(cc.URI, cc.Password, cc.Serial, cc.Interface, cc.Power, cc.Energy)
 }
 
 // NewSMA creates a SMA Meter
-func NewSMA(uri, serial, iface, power, energy string) (api.Meter, error) {
+func NewSMA(uri, password, serial, iface, power, energy string) (api.Meter, error) {
 	log := util.NewLogger("sma")
 
-	sm := &SMA{
-		mux:     util.NewWaiter(udpTimeout, func() { log.TRACE.Println("wait for initial value") }),
-		log:     log,
-		uri:     uri,
-		serial:  serial,
-		iface: iface,
-		powerO:  sma.Obis(power),
-		energyO: sma.Obis(energy),
-		recv:    make(chan sma.Telegram),
+	if password == "" {
+		password = "0000"
 	}
 
-	if sma.Instance == nil {
-		instance, err := sma.New(log, iface)
-		if err != nil {
-			return nil, err
-		}
-		sma.Instance = instance
+	err := sunny.SetMulticastInterface(iface)
```

Comment:
Iface can be nil or empty

Reply:
nil ist nicht möglich für einen string in go

aber für den leeren string haben ich eine Behandlung eingebaut